{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "import json\n",
    "from botocore.exceptions import ClientError\n",
    "import time\n",
    "import pandas as pd\n",
    "import datetime\n",
    "from datetime import datetime\n",
    "\n",
    "client = boto3.client('s3'\n",
    "                      ,aws_access_key_id = 'AKIAW5TFIVP5MY3DD44R'\n",
    "                      , aws_secret_access_key = '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'\n",
    "                     ) #low-level functional API\n",
    "\n",
    "resource = boto3.resource('s3') #high-level object-oriented API\n",
    "my_bucket = resource.Bucket('binance-macro-economy') #subsitute this for your s3 bucket name.\n",
    "\n",
    "# obj = client.get_object(Bucket='', Key='')\n",
    "session = boto3.Session(aws_access_key_id = 'AKIAW5TFIVP5MY3DD44R'\n",
    "                        ,aws_secret_access_key = '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'\n",
    "                        ,region_name = 'us-east-2')\n",
    "s3 = session.resource('s3')\n",
    "\n",
    "session_start_time = datetime.now()\n",
    "\n",
    "start_time = 1637202776 # 2021-11-18 02:32:56 (not sure of timezone)\n",
    "stop_time = int(time.time())\n",
    "# stop_time = 1637203000\n",
    "\n",
    "symbol = []\n",
    "price = []\n",
    "seconds = []\n",
    "for i in range(start_time,stop_time):\n",
    "    try:\n",
    "        s3.Object('binance-macro-economy', 's3_obj_{}.json'.format(i))\n",
    "        content_object = s3.Object('binance-macro-economy', 's3_obj_{}.json'.format(i))\n",
    "        file_content = content_object.get()['Body'].read()\n",
    "        my_json = json.loads(file_content.decode('utf8').replace(\"'\", '\"'))\n",
    "\n",
    "        for j in range(len(my_json)-1):\n",
    "            symbol.append(my_json[j]['symbol'])\n",
    "            price.append(my_json[j]['price'])\n",
    "            seconds.append(i)\n",
    "    except ClientError:\n",
    "        # Not found\n",
    "        pass\n",
    "\n",
    "df = pd.DataFrame([seconds,symbol,price]).T.rename(columns={0: 'seconds', 1: 'symbol', 2: 'price'})\n",
    "\n",
    "df_t = pd.pivot_table(df,index='seconds',columns='symbol',values='price').reset_index()\n",
    "\n",
    "df_t['timestamp'] = pd.to_datetime(df_t['seconds'], unit='s')\n",
    "\n",
    "df_t.to_csv('binance_macro_economy.csv', index = False)\n",
    "\n",
    "session_stop_time = datetime.now()\n",
    "session_total_time = session_stop_time - session_start_time\n",
    "\n",
    "print(session_total_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "InterfaceError",
     "evalue": "('IM002', '[IM002] [Microsoft][ODBC Driver Manager] Data source name not found and no default driver specified (0) (SQLDriverConnect)')",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mInterfaceError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_18164/3516737800.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mpyodbc\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mconn\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpyodbc\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mconnect\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'Driver={Amazon Redshift (x64)}; Server=redshift-cluster-1.csvogyduuzab.us-east-2.redshift.amazonaws.com; Database=dev'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mInterfaceError\u001b[0m: ('IM002', '[IM002] [Microsoft][ODBC Driver Manager] Data source name not found and no default driver specified (0) (SQLDriverConnect)')"
     ]
    }
   ],
   "source": [
    "import pyodbc \n",
    "Driver={Amazon Redshift (x64)}; Server=; Database=dev\n",
    "cnxn = pyodbc.connect('DRIVER={Devart ODBC Driver for Amazon Redshift};Server=redshift-cluster-1.csvogyduuzab.us-east-2.redshift.amazonaws.com;Database=dev;User ID=myuserid;Password=mypassword;Port=22;String Types=Unicode')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2\n",
    "import pprint\n",
    "\n",
    "\n",
    "configuration = { 'dbname': 'dev', \n",
    "                  'user':'user_name',\n",
    "                  'pwd':'user_password',\n",
    "                  'host':'redshift_endpoint',\n",
    "                  'port':'22'\n",
    "                }\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def create_conn(*args,**kwargs):\n",
    "\n",
    "    config = kwargs['config']\n",
    "    try:\n",
    "        conn=psycopg2.connect(dbname=config['dbname'], host=config['host'], port=config['port'], user=config['user'], password=config['pwd'])\n",
    "    except Exception as err:\n",
    "        print err.code, err\n",
    "\n",
    "\n",
    "    return conn\n",
    "\n",
    "\n",
    "def select(*args,**kwargs):\n",
    "    # need a connection with dbname=<username>_db\n",
    "    cur = kwargs['cur']\n",
    "\n",
    "    try:\n",
    "        # retrieving all tables in my search_path\n",
    "        cur.execute(\"\"\"select tablename from pg_table_def\"\"\")\n",
    "    except Exception as err:\n",
    "            print err.code,err\n",
    "\n",
    "\n",
    "    rows = cur.fetchall()\n",
    "    for row in rows:\n",
    "        print row\n",
    "\n",
    "\n",
    "print 'start'\n",
    "conn = create_conn(config=configuration)\n",
    "cursor = conn.cursor()\n",
    "print 'start select'\n",
    "select(cur=cursor)\n",
    "print 'finish'\n",
    "\n",
    "cursor.close()\n",
    "for n in conn.notices():\n",
    "    pprint(n)\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    2021-11-18 02:32:56\n",
       "1    2021-11-18 02:32:58\n",
       "2    2021-11-18 02:33:00\n",
       "3    2021-11-18 02:33:02\n",
       "4    2021-11-18 02:33:03\n",
       "             ...        \n",
       "84   2021-11-18 02:34:53\n",
       "85   2021-11-18 02:34:54\n",
       "86   2021-11-18 02:34:56\n",
       "87   2021-11-18 02:34:57\n",
       "88   2021-11-18 02:34:58\n",
       "Name: timestamp, Length: 89, dtype: datetime64[ns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "cannot convert the series to <class 'int'>",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_4244/495232501.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mdatetime\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mdatetime\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mdf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mt\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'date'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdatetime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfromtimestamp\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mt\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'seconds'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstrftime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"%A, %B %d, %Y %I:%M:%S\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\pandas\\core\\series.py\u001b[0m in \u001b[0;36mwrapper\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    183\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    184\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mconverter\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 185\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf\"cannot convert the series to {converter}\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    186\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    187\u001b[0m     \u001b[0mwrapper\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34mf\"__{converter.__name__}__\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: cannot convert the series to <class 'int'>"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "df.t['date'] = datetime.fromtimestamp(df.t['seconds']).strftime(\"%A, %B %d, %Y %I:%M:%S\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "type object 'datetime.datetime' has no attribute 'datetime'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_4244/1341836383.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mb\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0ma\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mdatetime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdatetime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtimedelta\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m3\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m: type object 'datetime.datetime' has no attribute 'datetime'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
